### PR TITLE
Feature – Edge bundling

### PR DESCRIPTION
**Pull request by simon-burke**
# Edge bundling
## Description
Adds edge bundling functionality to visualize connections between data points. Uses datashader's hammer edge bundling to convert edges into line segments, interpolates segment colors based on labels, and renders the segments in the datashader and matplotlib renderers.

## Features
- Integrate hammer edge bundling for static plots
- Add support for plotting lines in matplotlib and datashader renders
- Add test cases for edge bundling feature

## Basic Usage
```python
fig, ax = datamapplot.create_plot(
    datamap,
    labels,
    edge_bundle=True,
    edge_bundle_keywords={"color_map_nn": 100, "edges": edges},
    matplotlib_lineswidth=0.02
)
```

## Arguments
Arguments are passed into `datamapplot.create_plot()`
- `edge_bundle`: Boolean, determines whether or not to perform edge bundling
- `matplotlib_lineswidth`: Float, default=0.05, determines width of line segments when plotted in matplotlib
- `edge_bundling_kwargs`: dict, additional arguments for edge bundling
### edge_bundling_kwargs
- `edges=None`: dataframe, Pandas dataframe with source and target indices. If None, build edges using KNN.
- `color_map_nn=100`: Int, number of nearest neighbors to consider when interpolating segment colors
- `n_neighbors=10`: Int, number of nearest neighbors to connect if edges are not provided
- `hammer_bundle_kwargs={"use_dask":False}` dict, kwargs to pass to datashader's hammer bundle

## Examples
Arxiv ML articles (using KNN graph)
image
Supervised UMAP with Phylum taxonomy labels (using UMAP connectivity graph)
image

### CODE REVIEW

File: datamapplot/edge_bundling.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    edges=None,
+    hammer_bundle_kwargs=None,
+):
+    if hammer_bundle_kwargs is None:
+        hammer_bundle_kwargs = {"use_dask": False}
+
+    """Use hammer edge bundling on nearest neighbors"""
+    nbrs = NearestNeighbors(
+        n_neighbors=max(n_neighbors, color_map_nn), algorithm="ball_tree", n_jobs=-1
+    ).fit(data_map_coords)
```
Comment (lmcinnes) [inline datamapplot/edge_bundling.py:93]: I think we can move this under the ``if edges is None`` clause.

**Review by lmcinnes — CHANGES_REQUESTED** [review]
This looks great! I'm glad I put in some work on speeding up datashader's hammer edge bundling last year. One minor nit with respect to fitting the nearest neighbor model when you don't have to, but otherwise ready for merge (the test failures seem to be expected with a new comparison image).

File: datamapplot/edge_bundling.py (hunk 1 above)
Comment (simon-burke) [reply datamapplot/edge_bundling.py:93]: Your update to datashader's hammer bundle was one of the main reasons I decided to work on this. It's amazing to have a edge bundling implementation in python that is fast enough to use on larger datasets.

The nearest neighbors graph is used to interpolate line segment colors even when edges are provided. You can see this in the call to `_interpolate_colors(nbrs, color_list, midpoints, nn=color_map_nn) ` on line 129.

If you have another suggestion for how to color the line segments, I'm open to alternatives. I tried a few other options such as coloring based on the parent nodes, but found the smooth transition between colors with the nearest neighbors implementation to be much more inspiring.

**Comment by lmcinnes** [comment]
Thanks again for all your work on this. It is greatly appreciated!
